### PR TITLE
Implemented random sampling

### PR DESCRIPTION
**Pull request by beniwohli**
Given a sample rate, this will randomly choose transactions
to be sampled. If a transaction isn't sampled, no spans will
be collected, as well as no context information

### CODE REVIEW

File: elasticapm/traces.py [hunk 1]
```diff
@@ -230,14 +244,14 @@ def decorated(*args, **kwds):
     def __enter__(self):
         transaction = get_transaction()
 
-        if transaction:
+        if transaction and transaction.is_sampled:
             transaction.begin_span(self.name, self.type, self.extra,
```
Comment (roncohen) [inline elasticapm/traces.py]: Do you need the check for `in_sampled` it both here and inside `begin_span`?

File: elasticapm/instrumentation/packages/base.py [hunk 2]
```diff
@@ -219,7 +219,8 @@ def uninstrument(self):
         self.originals = {}
 
     def call_if_sampling(self, module, method, wrapped, instance, args, kwargs):
-        if not get_transaction():
+        transaction = get_transaction()
+        if not transaction or not transaction.is_sampled:
             return wrapped(*args, **kwargs)
```
Comment (roncohen) [inline elasticapm/instrumentation/packages/base.py]: you have the check in `begin_span`. Do you need it here too?

File: elasticapm/traces.py [hunk 3]
```diff
@@ -102,7 +112,8 @@ def to_dict(self):
             'context': self._context,
             'spans': [
```
Comment (roncohen) [inline elasticapm/traces.py]: I think it's worth considering removing the `spans` key entirely to clearly indicate this is an unsampled transaction.

File: elasticapm/instrumentation/packages/base.py (hunk 2 above)
Comment (beniwohli) [reply elasticapm/instrumentation/packages/base.py]: not 100% sure tbh. I'll step through it with the debugger

Comment (beniwohli) [reply elasticapm/instrumentation/packages/base.py]: After taking a closer read of the code, I think it's worth doing the check in both places.

`call_if_sampling` is called for stuff we instrument automatically. Checking here avoids even calling the `capture_span` decorator and doing other stuff we do in the wrappers (e.g. parsing the SQL query for an `sql` span).

the check in `capture_span.__enter__` is useful for cases where code is instrumented manually using the `capture_span` decorator.

The third check, in `Transaction.begin_span` seems to be superfluous, however, as it is only called from `capture_span.__enter__`. Good catch!

**Pull request closed without merging**